joomla/joomla-cms
# Subforms: Implementing per-field filters and validation.

This pull request introduces the option to both filter and validate subforms field by field. The intention is that every single field inside a subform can have its own filter and validation parameters. Before this PR, these parameters would just be ignored.

### Summary of Changes
* Introduced a Parameter `filter="subform"` to filter a subform field by filtering its child fields individually.
* Introduced a Parameter `validate="subform"` to validate a subform field by validating its child fields individually.
* Moved the loading of the JForm instance and its data inside the subform field from the `getInput` method into two separate methods to avoid repeated code, because the subform has to be loaded for filtering and validation, too. This change should **not** change any present behavior of subform fields, because I only moved the existing code into two additional methods.

### Testing Instructions
* Test with a form containing a subform field.
  * I created a small module for testing purposes on [mod_demovalidation](https://github.com/Harmageddon/mod_demovalidation). It has no frontend, only the form in the module configuration in backend. Feel free to use this, but I'd recommend that you play around with different parameter combinations to not just replicate my tests.
  * You can also use the testing instructions from #7829 and add some filters and validations...
  * ... or those from #12813, where an inline subform is used.
  * If you are developing an extension using subforms, this is the perfect opportunity to improve it! ;-)
* Test the impact of filters.
  * Add `filter="subform"` to the subform field.
  * For example, add `filter="int"` to a field inside the subform XML and try to save values like "abc" (should be filtered to 0) or "0002" (should be filtered to 2).
  * Make sure that correct values are not filtered and everything is stored correctly.
* Test the impact of validation.
  * Add `validate="subform"` to the subform field.
  * For example, add `validate="email"` to a field inside the subform XML and try to save values like "foo".
  * Make sure that correct values are not marked as invalid and everything is stored correctly.
* Make sure that the subform field behaves like before (apart from the added filtering and validation options).
* Test everything with `multiple="true"`, `multiple="false"` and without the `multiple` parameter for the subform field.

### Expected result
* Unfiltered values should be filtered, depending on the filters of the fields inside the subform XML.
* Invalid values should be rejected, depending on the validation rules of the fields inside the subform XML.

If a validation fails, there are three options what to display as an error message:
1. "Invalid Field: Subform" (Title of the subform field containing at least one invalid value)
2. "Invalid Field: Mail" (Title of the first invalid single field inside the subform)
3. Like 2., but for all invalid fields.

I used option 2, with a fallback to 1. For a subform containing invalid values, only one warning message will be returned, containing the title of the first field with an invalid value. I'm preferring this over option 1, because it is more specific about where the error is. Option 3 would be possible using additional messages, but as the basic implementation of validation itself displays a maximum of 3 messages, I thought it would be sufficient to send only one here.

### Documentation Changes Required
* ~As I have no experience in the version numbering of the CMS libraries, I'd be thankful if someone could tell me which value should be added for the `@since` tag for the `JFormRuleSubform` class.~
* When the pull request is merged and released, I will add short documentation about the additional parameters on docs.joomla.org.
